File: adapters/shared/android/src/main/java/com/sandrox/sfiora/bridge/SfioraBridgeOptions.java
package com.sandrox.sfiora.bridge;

import com.sandrox.sfiora.NfcReadConfiguration;
import com.sandrox.sfiora.NfcReadMode;
import com.sandrox.sfiora.ui.NfcScanPresentation;
import com.sandrox.sfiora.ui.NfcPresentationMessages;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Strict Android-side parser for the frozen bridge scan contract.
 */
public final class SfioraBridgeOptions {
    private static final Set<String> ROOT_KEYS = setOf(
            "mode",
            "timeoutMilliseconds",
            "android",
            "ios",
            "messages"
    );
    private static final Set<String> ANDROID_KEYS = setOf(
            "presentation",
            "deepReadEnabled",
            "presenceCheckDelayMilliseconds"
    );
    private static final Set<String> IOS_KEYS = setOf("pollingTechnologies");
    private static final Set<String> IOS_POLLING_TECHNOLOGIES = setOf(
            "iso14443",
            "iso15693",
            "iso18092"
    );

    private final NfcReadConfiguration readConfiguration;
    private final NfcScanPresentation presentation;
    private final NfcPresentationMessages.Scan messages;

    private SfioraBridgeOptions(
            NfcReadConfiguration readConfiguration,
            NfcScanPresentation presentation,
            NfcPresentationMessages.Scan messages
    ) {
        this.readConfiguration = readConfiguration;
        this.presentation = presentation;
        this.messages = messages;
    }

    public NfcReadConfiguration getReadConfiguration() {
        return readConfiguration;
    }

    public NfcScanPresentation getPresentation() {
        return presentation;
    }

    public NfcPresentationMessages.Scan getMessages() {
        return messages;
    }

    public static SfioraBridgeOptions parse(Object rawOptions)
            throws SfioraBridgeOptionsException {
        try {
            return parseValidated(rawOptions);
        } catch (IllegalArgumentException error) {
            throw new SfioraBridgeOptionsException(error.getMessage());
        }
    }

    private static SfioraBridgeOptions parseValidated(Object rawOptions)
            throws SfioraBridgeOptionsException {
        Map<?, ?> options = objectMap(rawOptions, "options", true);
        validateKeys(options, ROOT_KEYS, "options");

        NfcReadConfiguration.Builder configuration = NfcReadConfiguration.builder();
        if (options.containsKey("mode")) {
            configuration.mode(readMode(options.get("mode")));
        }
        if (options.containsKey("timeoutMilliseconds")) {
            configuration.timeoutMillis(integer(
                    options.get("timeoutMilliseconds"),
                    "timeoutMilliseconds"
            ));
        }

        NfcScanPresentation presentation = NfcScanPresentation.MANAGED;
        if (options.containsKey("android")) {
            Map<?, ?> androidOptions = objectMap(
                    options.get("android"),
                    "android",
                    false
            );
            validateKeys(androidOptions, ANDROID_KEYS, "android");
            if (androidOptions.containsKey("presentation")) {
                String value = string(
                        androidOptions.get("presentation"),
                        "android.presentation"
                );
                try {
                    presentation = NfcScanPresentation.fromBridgeValue(value);
                } catch (IllegalArgumentException error) {
                    throw new SfioraBridgeOptionsException(error.getMessage());
                }
            }
            if (androidOptions.containsKey("deepReadEnabled")) {
                configuration.deepReadEnabled(bool(
                        androidOptions.get("deepReadEnabled"),
                        "android.deepReadEnabled"
                ));
            }
            if (androidOptions.containsKey("presenceCheckDelayMilliseconds")) {
                long delay = integer(
                        androidOptions.get("presenceCheckDelayMilliseconds"),
                        "android.presenceCheckDelayMilliseconds"
                );
                if (delay < Integer.MIN_VALUE || delay > Integer.MAX_VALUE) {
                    throw new SfioraBridgeOptionsException(
                            "android.presenceCheckDelayMilliseconds is outside the integer range"
                    );
                }
                configuration.presenceCheckDelayMillis((int) delay);
            }
        }

        if (options.containsKey("ios")) {
            validateIosOptions(options.get("ios"));
        }

        try {
            NfcPresentationMessages.Scan messages = options.containsKey("messages")
                    ? SfioraBridgePresentationMessages.parseScan(options.get("messages"))
                    : null;
            return new SfioraBridgeOptions(
                    configuration.build(),
                    presentation,
                    messages
            );
        } catch (IllegalArgumentException error) {
            throw new SfioraBridgeOptionsException(error.getMessage());
        }
    }

    private static NfcReadMode readMode(Object value)
            throws SfioraBridgeOptionsException {
        String mode = string(value, "mode");
        switch (mode) {
            case "automatic":
                return NfcReadMode.AUTOMATIC;
            case "ndef":
                return NfcReadMode.NDEF;
            case "discover":
                return NfcReadMode.DISCOVER;
            default:
                throw new SfioraBridgeOptionsException(
                        "mode must be automatic, ndef, or discover"
                );
        }
    }

    private static void validateIosOptions(Object value)
            throws SfioraBridgeOptionsException {
        Map<?, ?> iosOptions = objectMap(value, "ios", false);
        validateKeys(iosOptions, IOS_KEYS, "ios");
        if (!iosOptions.containsKey("pollingTechnologies")) {
            return;
        }
        Object technologiesValue = iosOptions.get("pollingTechnologies");
        if (!(technologiesValue instanceof List<?>)) {
            throw new SfioraBridgeOptionsException(
                    "ios.pollingTechnologies must be an array"
            );
        }
        List<?> technologies = (List<?>) technologiesValue;
        if (technologies.isEmpty()) {
            throw new SfioraBridgeOptionsException(
                    "ios.pollingTechnologies must not be empty"
            );
        }
        Set<String> unique = new HashSet<>();
        for (int index = 0; index < technologies.size(); index++) {
            String technology = string(
                    technologies.get(index),
                    "ios.pollingTechnologies[" + index + "]"
            );
            if (!IOS_POLLING_TECHNOLOGIES.contains(technology)) {
                throw new SfioraBridgeOptionsException(
                        "ios.pollingTechnologies contains an unsupported value"
                );
            }
            if (!unique.add(technology)) {
                throw new SfioraBridgeOptionsException(
                        "ios.pollingTechnologies must not contain duplicates"
                );
            }
        }
    }

    private static Map<?, ?> objectMap(
            Object value,
            String name,
            boolean nullAsEmpty
    ) throws SfioraBridgeOptionsException {
        if (value == null && nullAsEmpty) {
            return Collections.emptyMap();
        }
        if (!(value instanceof Map<?, ?>)) {
            throw new SfioraBridgeOptionsException(name + " must be an object");
        }
        return (Map<?, ?>) value;
    }

    private static void validateKeys(
            Map<?, ?> values,
            Set<String> allowedKeys,
            String name
    ) throws SfioraBridgeOptionsException {
        for (Object rawKey : values.keySet()) {
            if (!(rawKey instanceof String) || !allowedKeys.contains(rawKey)) {
                throw new SfioraBridgeOptionsException(
                        name + " contains an unsupported field: " + rawKey
                );
            }
        }
    }

    private static String string(Object value, String name)
            throws SfioraBridgeOptionsException {
        if (!(value instanceof String) || ((String) value).isEmpty()) {
            throw new SfioraBridgeOptionsException(name + " must be a non-empty string");
        }
        return (String) value;
    }

    private static boolean bool(Object value, String name)
            throws SfioraBridgeOptionsException {
        if (!(value instanceof Boolean)) {
            throw new SfioraBridgeOptionsException(name + " must be a boolean");
        }
        return (Boolean) value;
    }

    private static long integer(Object value, String name)
            throws SfioraBridgeOptionsException {
        if (!(value instanceof Number)) {
            throw new SfioraBridgeOptionsException(name + " must be an integer");
        }
        double number = ((Number) value).doubleValue();
        if (!Double.isFinite(number)
                || Math.rint(number) != number
                || number < Long.MIN_VALUE
                || number > Long.MAX_VALUE) {
            throw new SfioraBridgeOptionsException(name + " must be an integer");
        }
        return (long) number;
    }

    private static Set<String> setOf(String... values) {
        return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(values)));
    }
}

File: adapters/shared/ios/Sources/SfioraBridgeSupport/SfioraBridgeOptions.swift
import CoreFoundation
import Foundation

#if canImport(Sfiora)
    import Sfiora
#endif

enum SfioraBridgeOptionsError: Error, Equatable, LocalizedError {
    case invalid(String)

    var errorDescription: String? {
        switch self {
        case .invalid(let message):
            return message
        }
    }
}

struct SfioraBridgeOptions {
    private static let rootKeys: Set<String> = [
        "mode",
        "timeoutMilliseconds",
        "android",
        "ios",
        "messages",
    ]
    private static let androidKeys: Set<String> = [
        "presentation",
        "deepReadEnabled",
        "presenceCheckDelayMilliseconds",
    ]
    private static let iosKeys: Set<String> = ["pollingTechnologies"]
    private static let pollingTechnologyValues: Set<String> = [
        "iso14443",
        "iso15693",
        "iso18092",
    ]

    let readerConfiguration: NfcReadConfiguration

    static func parse(_ rawOptions: Any?) throws -> SfioraBridgeOptions {
        let options = try dictionary(
            rawOptions,
            name: "options",
            nilAsEmpty: true
        )
        try validateKeys(options, allowed: rootKeys, name: "options")

        let mode = try options["mode"].map(readMode) ?? .automatic
        let timeoutMilliseconds =
            try options["timeoutMilliseconds"].map {
                try integer($0, name: "timeoutMilliseconds")
            } ?? 30_000

        if let rawAndroid = options["android"] {
            try validateAndroidOptions(rawAndroid)
        }

        let pollingTechnologies: Set<NfcPollingTechnology>
        if let rawIos = options["ios"] {
            pollingTechnologies = try readIosOptions(rawIos)
        } else {
            pollingTechnologies = [.iso14443, .iso15693]
        }
        let presentationMessages: NfcReaderPresentationMessages
        do {
            presentationMessages =
                try options["messages"].map(
                    SfioraBridgePresentationMessages.reader
                ) ?? .standard
        } catch {
            throw SfioraBridgeOptionsError.invalid(error.localizedDescription)
        }

        do {
            return SfioraBridgeOptions(
                readerConfiguration: try NfcReadConfiguration(
                    mode: mode,
                    timeoutMilliseconds: timeoutMilliseconds,
                    pollingTechnologies: pollingTechnologies,
                    presentationMessages: presentationMessages
                )
            )
        } catch let error as NfcConfigurationError {
            throw SfioraBridgeOptionsError.invalid(
                error.errorDescription ?? "The NFC scan options are invalid"
            )
        }
    }

    private static func readMode(_ value: Any) throws -> NfcReadMode {
        switch try string(value, name: "mode") {
        case "automatic":
            return .automatic
        case "ndef":
            return .ndef
        case "discover":
            return .discover
        default:
            throw SfioraBridgeOptionsError.invalid(
                "mode must be automatic, ndef, or discover"
            )
        }
    }

    private static func validateAndroidOptions(_ value: Any) throws {
        let options = try dictionary(value, name: "android")
        try validateKeys(options, allowed: androidKeys, name: "android")

        if let presentation = options["presentation"] {
            let value = try string(
                presentation,
                name: "android.presentation"
            )
            guard value == "managed" || value == "none" else {
                throw SfioraBridgeOptionsError.invalid(
                    "android.presentation must be managed or none"
                )
            }
        }

        if let deepReadEnabled = options["deepReadEnabled"] {
            try boolean(
                deepReadEnabled,
                name: "android.deepReadEnabled"
            )
        }

        if let rawDelay = options["presenceCheckDelayMilliseconds"] {
            let delay = try integer(
                rawDelay,
                name: "android.presenceCheckDelayMilliseconds"
            )
            guard 50...5_000 ~= delay else {
                throw SfioraBridgeOptionsError.invalid(
                    "android.presenceCheckDelayMilliseconds must be between 50 and 5000"
                )
            }
        }
    }

    private static func readIosOptions(
        _ value: Any
    ) throws -> Set<NfcPollingTechnology> {
        let options = try dictionary(value, name: "ios")
        try validateKeys(options, allowed: iosKeys, name: "ios")

        guard let rawTechnologies = options["pollingTechnologies"] else {
            return [.iso14443, .iso15693]
        }
        guard let values = rawTechnologies as? [Any], !values.isEmpty else {
            throw SfioraBridgeOptionsError.invalid(
                "ios.pollingTechnologies must be a non-empty array"
            )
        }

        var rawUniqueValues = Set<String>()
        var technologies = Set<NfcPollingTechnology>()
        for (index, rawValue) in values.enumerated() {
            let name = "ios.pollingTechnologies[\(index)]"
            let value = try string(rawValue, name: name)
            guard pollingTechnologyValues.contains(value) else {
                throw SfioraBridgeOptionsError.invalid(
                    "ios.pollingTechnologies contains an unsupported value"
                )
            }
            guard rawUniqueValues.insert(value).inserted else {
                throw SfioraBridgeOptionsError.invalid(
                    "ios.pollingTechnologies must not contain duplicates"
                )
            }
            guard let technology = NfcPollingTechnology(rawValue: value) else {
                throw SfioraBridgeOptionsError.invalid(
                    "ios.pollingTechnologies contains an unsupported value"
                )
            }
            technologies.insert(technology)
        }
        return technologies
    }

    private static func dictionary(
        _ value: Any?,
        name: String,
        nilAsEmpty: Bool = false
    ) throws -> NSDictionary {
        if value == nil, nilAsEmpty {
            return NSDictionary()
        }
        guard let value = value as? NSDictionary else {
            throw SfioraBridgeOptionsError.invalid("\(name) must be an object")
        }
        return value
    }

    private static func validateKeys(
        _ values: NSDictionary,
        allowed: Set<String>,
        name: String
    ) throws {
        for rawKey in values.allKeys {
            guard
                let key = rawKey as? String,
                allowed.contains(key)
            else {
                throw SfioraBridgeOptionsError.invalid(
                    "\(name) contains an unsupported field: \(rawKey)"
                )
            }
        }
    }

    private static func string(_ value: Any, name: String) throws -> String {
        guard let value = value as? String, !value.isEmpty else {
            throw SfioraBridgeOptionsError.invalid(
                "\(name) must be a non-empty string"
            )
        }
        return value
    }

    @discardableResult
    private static func boolean(_ value: Any, name: String) throws -> Bool {
        guard
            let number = value as? NSNumber,
            CFGetTypeID(number) == CFBooleanGetTypeID()
        else {
            throw SfioraBridgeOptionsError.invalid("\(name) must be a boolean")
        }
        return number.boolValue
    }

    private static func integer(_ value: Any, name: String) throws -> Int {
        guard
            let number = value as? NSNumber,
            CFGetTypeID(number) != CFBooleanGetTypeID()
        else {
            throw SfioraBridgeOptionsError.invalid("\(name) must be an integer")
        }

        let doubleValue = number.doubleValue
        guard
            doubleValue.isFinite,
            doubleValue.rounded(.towardZero) == doubleValue,
            doubleValue >= Double(Int.min),
            doubleValue <= Double(Int.max)
        else {
            throw SfioraBridgeOptionsError.invalid("\(name) must be an integer")
        }
        return Int(doubleValue)
    }
}
